Add basic GitHub Actions CI

runfalk opened the pull request:
Hey there, really nice project! I saw that there is no CI set up so I did a basic config. I did an opionated one with formatting and clippy linting. Let me know what you like and dislike.

Related to #29.

andrewchambers commented:
Looks pretty good, I would maybe shift the pre-commit hook into the support directory. I also wonder what additional platforms we can test on.

I might tweak some of the error messages in the pre-commit hook, though nothing too major.

andrewchambers commented:
I definitely should refactor the code to avoid the need for those exceptions too haha.

andrewchambers commented on pre-commit:
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
Could we split the commit hook into a different pull request?

runfalk replied:
Done!

runfalk commented:
So, I've split the pre-commit hook to a separate PR.

I've also updated this one to test on both macOS and Ubuntu latest. It doesn't seem like GitHub Actions support FreeBSD or ARM (unless you use your own runners perhaps https://medium.com/@nabil.servais/using-github-actions-to-test-your-go-code-on-arm64-ce581e646cb).

I noticed that the `bats` suite actually fails on macOS (at least when run  on GitHub actions):

```
1..33
ok 1 simple put/get primary key
ok 2 simple put/get put key
ok 3 simple put/get no compression
not ok 4 random data
# (in test file cli-tests/cli-tests.bats, line 77)
#   `head -c $i /dev/urandom > "$SCRATCH/rand.dat"' failed
# head: illegal byte count -- 0
not ok 5 highly compressible data
# (in test file cli-tests/cli-tests.bats, line 89)
#   `yes | head -c $i > "$SCRATCH/yes.dat"' failed
# head: illegal byte count -- 0
ok 6 key mismatch
ok 7 corruption detected
ok 8 concurrent send
ok 9 simple search and listing
not ok 10 rm and gc
# (in test file cli-tests/cli-tests.bats, line 177)
#   `test 2 = "$(ls "$BUPSTASH_REPOSITORY"/data | wc -l)"' failed
not ok 11 rm and restore-removed
# (in test file cli-tests/cli-tests.bats, line 214)
#   `test 2 = "$(bupstash list | wc -l)"' failed
ok 12 query sync
ok 13 get via query
ok 14 rm via query
not ok 15 send directory sanity
# (in test file cli-tests/cli-tests.bats, line 300)
#   `test 5 = "$(bupstash get id=$id | tar -tf - | wc -l)"' failed
not ok 16 send directory no stat cache
# (in test file cli-tests/cli-tests.bats, line 313)
#   `test 5 = "$(bupstash get id=$id | tar -tf - | wc -l)"' failed
ok 17 stat cache invalidated
ok 18 repository command
ok 19 key command
not ok 20 long path
# (in test file cli-tests/cli-tests.bats, line 358)
#   `test 7 = "$(bupstash get id=$id | tar -tf - | wc -l)"' failed
#        7
not ok 21 long link target
# (in test file cli-tests/cli-tests.bats, line 370)
#   `test 2 = "$(bupstash get id=$id | tar -tf - | wc -l)"' failed
not ok 22 directory exclusions
# (in test file cli-tests/cli-tests.bats, line 380)
#   `test 4 = "$(bupstash get id=$id | tar -tf - | wc -l)"' failed
ok 23 checkpoint plain data
not ok 24 checkpoint directories
# (in test file cli-tests/cli-tests.bats, line 413)
#   `test 4 = "$(bupstash get id=$id | tar -tf - | wc -l)"' failed
not ok 25 rm from stdin
# (in test file cli-tests/cli-tests.bats, line 420)
#   `test 3 = "$(bupstash list | wc -l)"' failed
ok 26 concurrent dir modify during put
not ok 27 list and rm no key
# (in test file cli-tests/cli-tests.bats, line 479)
#   `test 2 = "$(bupstash list --query-encrypted | wc -l)"' failed
# cfecc985f05eb87804253ad01b2ba638
# 57e3ffaabf126c5dc759dc6ae92080be
not ok 28 pick and index
# (in test file cli-tests/cli-tests.bats, line 493)
#   `head -c "$(shuf -i 10-1000 -n 1)" /dev/urandom > "$SCRATCH/foo/$(uuidgen)"' failed
# /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/bats.1739.src: line 493: shuf: command not found
# head: illegal byte count -- 
not ok 29 multi dir put
# (in test file cli-tests/cli-tests.bats, line 524)
#   `test 5 = "$(bupstash get id=$id | tar -tf - | wc -l)"' failed
# .
# bar
# bar/baz
# bar/baz/a.txt
# bang
ok 30 hard link short path
ok 31 long hard link target
not ok 32 hard link to symlink
# (in test file cli-tests/cli-tests.bats, line 566)
#   `readlink "$SCRATCH/restore/c"' failed
# x .
# x a
# x b
# x c
ok 33 # skip (Set PICK_TORTURE_DIR to run this test.) pick torture
Error: The operation was canceled.
```

It gets stuck after this and the run has to be cancelled or timed out.

andrewchambers commented:
> head: illegal byte count -- 0

Strange laws they have over there in macos.  Thanks for trying that, looks like I need to do some investigation.

runfalk commented:
Yeah, those failing tests may in part be due to that macOS doesn't ship with the GNU core utils.

`shuf` also seems to be missing. I found this Stack exchange discussion: https://apple.stackexchange.com/questions/142860/install-shuf-on-os-x. Do you prefer to modify the tests to not use the GNU specific behavior or to install GNU `coreutils` (might require some trickery with the script since they are prefixed with g by default)?

andrewchambers commented:
> Do you prefer to modify the tests to not use the GNU specific behavior or to install GNU coreutils (might require some trickery with the script since they are prefixed with g by default)?

Not totally decided yet, i think if they are failing we can go ahead and disable macos CI for now and get this up and running, then re enable after we had some time to think about the issues.

Sorry I'm being quite slow on this, i've just been a bit busy on other tasks.

runfalk commented:
I disabled macOS for the bats suite but left the `cargo test --release` in.

There is a apparently a bug with the caching action and macOS's BSD tar implementation, so I disabled all caching for the macOS tests which adds about a minute to that job. I think the CLI run is always slower though so shouldn't matter too much.

I'm pretty happy with the state of the PR as it is right now. Of course there could be more improvements, like automatically building the statically linked MUSL binary for all platforms, running torture tests and running the cli tests against a remote repository over SSH.

I did some more thinking about using `coreutils` stuff in the bats suite. It's probably a bad idea to use the unprefixed versions of the GNU commands since `bupstash` will likely be used with the system `tar` implementation. One option would be a variable like `SYS_CMD_PREFIX` that gets set to `g` on macOS and then say that we require `coreutils` to run the test suite, but not the actual application. I think I'm slightly leaning towards using the macOS defaults since the developer experience would be slightly nicer.

The use of `head -c 0 > /dev/urandom > $SCRATCH/rand.dat` could be replaced by `echo -n > $SCRATCH/rand.dat`. And actually the `random data` and `highly compressible data` test actually does the exact same thing when `i` is 0, so we only need the workaround for one of them.

The use of `shuf` could be replaced by:

```
# Small files (10-1,009 bytes)
$((10 + $(head -c 4 /dev/urandom | cksum | cut -f1 -d " " | head -c 3)))

# Large files (10,000 - 10,009,999 bytes)
$((10000 + $(head -c 4 /dev/urandom | cksum | cut -f1 -d " " | head -c 7)))
```

Which is close enough to the intended sizes I think, though it's a bit clunky. Getting random numbers in a cross platform way seems pretty difficult. On Linux `jot` only changes value every second, and `$RANDOM` is not available on macOS since it ships with Bash v3.

andrewchambers commented:
Thank you so much!
